### PR TITLE
feat: dyslexia_words_swap

### PR DESCRIPTION
**Pull request by notricenotsweet**
Adding dyslexia_words_swap transformation.

This transformation acts like a perturbation by altering some words with mistakes that are likely to happen in the context of dyslexia.

### CODE REVIEW

File: transformations/dyslexia_words_swap/README.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Dislexia Words Swap
+This transformation acts like a perturbation by altering some words with mistakes that are likely to happen in the context of dyslexia.
```
Comment (kaustubhdhole) [inline transformations/dyslexia_words_swap/README.md:2]: Please add your name/email/affiliation

Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/README.md:2]: Hello Kaustubh, cf my mail

File: transformations/dyslexia_words_swap/transformation.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class DyslexiaWordsSwap(SentenceOperation):
+
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+
+    def __init__(self):
```
Comment (csinva) [inline transformations/dyslexia_words_swap/transformation.py:13]: I believe the constructor should also take 'max_outputs' as one of the params, based on which generate function should return perturbed sentences.

File: transformations/dyslexia_words_swap/test.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (csinva) [inline transformations/dyslexia_words_swap/test.json:1]: Could you add a few more test cases here, particularly to see how this transformation deals with whitespace?

**Review by csinva — COMMENTED** [review]
Hi @**[notricenotsweet](https://github.com/notricenotsweet)**, thanks for this interesting contribution!

I've left some comments in the code, please take a look for some potential suggestions. Here's my general review:
**Correctness**: All checks have passed - please pull "main" once into your branch.
**Interface**: I think you've chosen the correct interface.
**Specificity**: The change is specific to anything which takes in text-level inputs.
**Novelty**: I should not there are two similar submitted transformations: [replace_spelling](https://github.com/GEM-benchmark/NL-Augmenter/tree/main/transformations/replace_spelling) randomly replaces  words with common misspellings and [correct_common_misspellings](https://github.com/GEM-benchmark/NL-Augmenter/pull/213) corrects the spelling of misspelled words in a text.  This transformation is very similar, but maybe different enough since it focuses on dyslexia words and swaps words in both directions (i.e. correct -> incorrect and incorrect -> correct).

**Adding New Libraries**: Nothing new added
**Test Cases**: 1 test cases added (should increase this somewhat)
**Evaluating Robustness**: Robustness evaluation is not yet conducted.

**Comment by mille-s** [comment]
Hi, thanks for the submission! I think it is a very nice idea, there are quite some works on addressing dyslexia in NLP, so it is also very relevant. I have one doubt though: the list of words to be replaced looks very short, is there any way to get a more extensive list? Otherwise I'm afraid the changes will only affect a few sentences...

File: transformations/dyslexia_words_swap/transformation.py (hunk 2 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/transformation.py]: Done. (even if we cannot generate more than one output, at least it will be consistent with the other constructors)

File: transformations/dyslexia_words_swap/test.json (hunk 3 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/test.json:1]: nothing particular with whitespace, only working with words. I am adding a few tests.

**Comment by notricenotsweet** [comment]
> Hi, thanks for the submission! I think it is a very nice idea, there are quite some works on addressing dyslexia in NLP, so it is also very relevant. I have one doubt though: the list of words to be replaced looks very short, is there any way to get a more extensive list? Otherwise I'm afraid the changes will only affect a few sentences...

The list encompasses the most frequent errors according to educators, and it seems words that will likely occur in the dataset, e.g. than, for, no, will. Do you have the total number of modification it will produce? If it is too few, we could try to complete the list, but the additional words will probably be less frequent ones, hence producing less modifications.

**Review by csinva — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
Seems you've copy pasted many files from earlier commits. The file difference shows 180 files. You will need to fix that so that future tracking is easy.

**Comment by notricenotsweet** [comment]
> Seems you've copy pasted many files from earlier commits. The file difference shows 180 files. You will need to fix that so that future tracking is easy.

Sorry for this, that should be ok now !

File: transformations/dyslexia_words_swap/transformation.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        for word in self.nlp(sentence):
+            new_sentence += sentence[end_idx: word.idx]
+
+            key = word.text.lower()
+            if key in self.swap_words or key in self.swap_words_2:
+                if key in self.swap_words:
+                    new_word = self.swap_words[key]
+                if key in self.swap_words_2:
+                    new_word = self.swap_words_2[key]
+                new_sentence += new_word
```
Comment (aadesh11) [inline transformations/dyslexia_words_swap/transformation.py:35]: 'new_word' might be referenced before assignment as it's getting assigned in the 'if' block.

File: transformations/dyslexia_words_swap/transformation.py [hunk 5]
```diff
@@ -0,0 +1,42 @@
+import os
+import json
+import spacy
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class DyslexiaWordsSwap(SentenceOperation):
```
Comment (aadesh11) [inline transformations/dyslexia_words_swap/transformation.py:8]: Please add docstrings for class and all the functions.

File: transformations/dyslexia_words_swap/transformation.py [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+    keywords = ["lexical", "rule-based", "external-knowledge-based", "aural", "possible-meaning-alteration", "low-precision", "low-coverage", "low-generations"]
+
+    def __init__(self, seed=0, max_outputs=1):
+        super().__init__(seed=seed, max_outputs=max_outputs)
+
+        self.nlp = spacy.load("en_core_web_sm")
```
Comment (aadesh11) [inline transformations/dyslexia_words_swap/transformation.py:17]: Please use global spacy instead of loading a new one. 

self.nlp = spacy_nlp if spacy_nlp else spacy.load("en_core_web_sm")

Refer to other transformations for more detail.

File: transformations/dyslexia_words_swap/transformation.py [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import spacy
+from interfaces.SentenceOperation import SentenceOperation
+from tasks.TaskTypes import TaskType
+
+
+class DyslexiaWordsSwap(SentenceOperation):
+
+    tasks = [TaskType.TEXT_CLASSIFICATION, TaskType.TEXT_TO_TEXT_GENERATION]
+    languages = ["en"]
+    keywords = ["lexical", "rule-based", "external-knowledge-based", "aural", "possible-meaning-alteration", "low-precision", "low-coverage", "low-generations"]
```
Comment (aadesh11) [inline transformations/dyslexia_words_swap/transformation.py:12]: Instead of a single line, write it in multiple lines.

File: transformations/dyslexia_words_swap/transformation.py (hunk 4 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/transformation.py]: Fixed ! Thanks !

File: transformations/dyslexia_words_swap/transformation.py (hunk 5 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/transformation.py:9]: Fixed ! Thanks !

File: transformations/dyslexia_words_swap/transformation.py (hunk 6 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/transformation.py]: Fixed ! Thanks !

File: transformations/dyslexia_words_swap/transformation.py (hunk 7 above)
Comment (notricenotsweet) [reply transformations/dyslexia_words_swap/transformation.py]: Fixed ! Thanks !

**Review by mille-s — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
Thanks for the effort @notricenotsweet and thank you @csinva, @mille-s and @aadesh11 for the review.

**Review by kaustubhdhole — APPROVED** [review]

**Comment by kaustubhdhole** [comment]
@notricenotsweet please send your name/email/affiliation again.
